File: java/Freeze/library/BookFactory.java
// **********************************************************************
//
// Copyright (c) 2003-2015 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

class BookFactory implements Ice.ObjectFactory
{
    @Override
    public Ice.Object
    create(String type)
    {
        assert(type.equals("::Demo::Book"));
        return new BookI(_library);
    }

    @Override
    public void
    destroy()
    {
    }

    BookFactory(LibraryI library)
    {
        _library = library;
    }

    private LibraryI _library;
}

File: cpp/Freeze/library/BookFactory.cpp
// **********************************************************************
//
// Copyright (c) 2003-2015 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <BookFactory.h>

using namespace std;

BookFactory::BookFactory(const LibraryIPtr& library) :
    _library(library)
{
}

Ice::ObjectPtr
#ifndef NDEBUG
BookFactory::create(const string& type)
#else
BookFactory::create(const string&)
#endif
{
    assert(_library);
    assert(type == "::Demo::Book");
    return new BookI(_library);
}

void
BookFactory::destroy()
{
    //
    // Break cyclic object dependencies
    //
    _library = 0;
}
